Array insert with reference type results in C error

**V version:** V 0.1.24 ed55778
**OS:** macOS 10.15

**What did you do?**

```v
$ cat f.v
struct Fail {
	name string
}

struct Fun {
mut:
	fails []&Fail
}

fn main() {
	f := Fail{}
	mut u := Fun{}
	u.fails << f

}
```

**What did you expect to see?**

Compile time error, suggesting to prefix 'f' with '&' like this:

```
- u.fails << f
+ u.fails << &f
```
 
**What did you see instead?**
```
$ v build -cg f.v
C compiler=cc
/private/var/folders/12/517hk27n0pxbnbv9yxlh4jxm0000gn/T/v/f.tmp.c:4217:68: error: initializing 'Fail *' (aka 'struct Fail *') with an expression of incompatible type 'Fail' (aka 'struct Fail'); take the address with &
_PUSH(& u .fails , ( /*typ = array_ptr_Fail   tmp_typ=Fail**/ f ), tmp1, Fail*) ;
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/private/var/folders/12/517hk27n0pxbnbv9yxlh4jxm0000gn/T/v/f.tmp.c:173:48: note: expanded from macro '_PUSH'
#define _PUSH(arr, val, tmp, tmp_typ) {tmp_typ tmp = (val); array_push(arr, &tmp);}
                                               ^     ~~~~~
1 error generated.
V error: C error. This should never happen.
Please create a GitHub issue: https://github.com/vlang/v/issues/new/choose
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1586,20 +1586,16 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				}
 				// []T << T or []T << []T
 				unwrapped_right_type := c.unwrap_generic(right_type)
-				mut hint := ''
-
 				if c.check_types(unwrapped_right_type, left_value_type) {
 					// []&T << T is wrong: we check for that, !(T.is_ptr()) && ?(&T).is_ptr()
-					if !unwrapped_right_type.is_ptr() && left_value_type.is_ptr()
-						&& left_value_type.share() == .mut_t {
-						hint = '. you can prefix the right argument with &.'
-					} else {
+					if unwrapped_right_type.is_ptr() || left_value_type.is_ptr()
+						|| left_value_type.share() == .mut_t {
 						return ast.void_type
 					}
 				} else if c.check_types(unwrapped_right_type, c.unwrap_generic(left_type)) {
 					return ast.void_type
 				}
-				c.error('cannot append `$right_sym.name` to `$left_sym.name`$hint', right_pos)
+				c.error('cannot append `$right_sym.name` to `$left_sym.name`', right_pos)
 				return ast.void_type
 			} else {
 				return c.check_shift(left_type, right_type, left_pos, right_pos)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1588,8 +1588,8 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				unwrapped_right_type := c.unwrap_generic(right_type)
 				if c.check_types(unwrapped_right_type, left_value_type) {
 					// []&T << T is wrong: we check for that, !(T.is_ptr()) && ?(&T).is_ptr()
-					if unwrapped_right_type.is_ptr() || left_value_type.is_ptr()
-						|| left_value_type.share() == .mut_t {
+					if !(!unwrapped_right_type.is_ptr() && left_value_type.is_ptr()
+						&& left_value_type.share() == .mut_t) {
 						return ast.void_type
 					}
 				} else if c.check_types(unwrapped_right_type, c.unwrap_generic(left_type)) {

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1586,12 +1586,13 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				}
 				// []T << T or []T << []T
 				unwrapped_right_type := c.unwrap_generic(right_type)
-				mut hint := ""
+				mut hint := ''
 
 				if c.check_types(unwrapped_right_type, left_value_type) {
 					// []&T << T is wrong: we check for that, !(T.is_ptr()) && ?(&T).is_ptr()
-					if !unwrapped_right_type.is_ptr() && left_value_type.is_ptr() && left_value_type.share() == .mut_t {
-						hint = ". you can prefix the right argument with &."
+					if !unwrapped_right_type.is_ptr() && left_value_type.is_ptr()
+						&& left_value_type.share() == .mut_t {
+						hint = '. you can prefix the right argument with &.'
 					} else {
 						return ast.void_type
 					}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -1586,11 +1586,19 @@ pub fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 				}
 				// []T << T or []T << []T
 				unwrapped_right_type := c.unwrap_generic(right_type)
-				if c.check_types(unwrapped_right_type, left_value_type)
-					|| c.check_types(unwrapped_right_type, c.unwrap_generic(left_type)) {
+				mut hint := ""
+
+				if c.check_types(unwrapped_right_type, left_value_type) {
+					// []&T << T is wrong: we check for that, !(T.is_ptr()) && ?(&T).is_ptr()
+					if !unwrapped_right_type.is_ptr() && left_value_type.is_ptr() && left_value_type.share() == .mut_t {
+						hint = ". you can prefix the right argument with &."
+					} else {
+						return ast.void_type
+					}
+				} else if c.check_types(unwrapped_right_type, c.unwrap_generic(left_type)) {
 					return ast.void_type
 				}
-				c.error('cannot append `$right_sym.name` to `$left_sym.name`', right_pos)
+				c.error('cannot append `$right_sym.name` to `$left_sym.name`$hint', right_pos)
 				return ast.void_type
 			} else {
 				return c.check_shift(left_type, right_type, left_pos, right_pos)

--- a/vlib/v/checker/tests/invalid_insert_references_test.out
+++ b/vlib/v/checker/tests/invalid_insert_references_test.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/invalid_insert_references_test.vv:8:7: error: cannot append `int literal` to `[]&int`. you can prefix the right argument with &.
+    6 |     a << &c
+    7 |     c = 2
+    8 |     a << 1
+      |          ^
+    9 |     println(a)
+   10 | }

--- a/vlib/v/checker/tests/invalid_insert_references_test.out
+++ b/vlib/v/checker/tests/invalid_insert_references_test.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/invalid_insert_references_test.vv:8:7: error: cannot append `int literal` to `[]&int`. you can prefix the right argument with &.
+vlib/v/checker/tests/invalid_insert_references_test.vv:8:7: error: cannot append `int literal` to `[]&int`
     6 |     a << &c
     7 |     c = 2
     8 |     a << 1

--- a/vlib/v/checker/tests/invalid_insert_references_test.vv
+++ b/vlib/v/checker/tests/invalid_insert_references_test.vv
@@ -1,0 +1,14 @@
+// fixes https://github.com/vlang/v/issues/3600, test based on a simplified version of example by https://github.com/radare
+fn test_invalid_insert_references() {
+	b := 0
+	mut a := [&b]
+	mut c := 1
+	a << &c
+	c = 2
+	a << 1
+	println(a)
+}
+
+fn main() {
+	test_invalid_insert_references()
+}

